rust/lance-linalg/build.rs

```diff
@@ -23,36 +23,59 @@ fn main() {
 
     println!("cargo:rerun-if-changed=src/simd/f16.c");
 
-    if cfg!(all(target_os = "macos", target_arch = "aarch64")) {
-        cc::Build::new()
-            .compiler("clang")
-            .file("src/simd/f16.c")
-            .flag("-mtune=apple-m1")
-            .flag("-O3")
-            .flag("-Wall")
-            .flag("-Werror")
-            .flag("-Wextra")
-            .flag("-Wpedantic")
-            .compile("f16");
+    if cfg!(any(not(feature = "fp16kernels"), target_os = "windows")) {
+        // We only compile the f16 kernels if the feature is enabled
+        // MSVC does not support the f16 type, so we also skip it on Windows.
+        return;
     }
 
-    if cfg!(all(target_os = "linux", feature = "avx512fp16")) {
-        // No fp16 without AVX512fp16
-        cc::Build::new()
-            .compiler("clang")
-            .std("c17")
-            .file("src/simd/f16.c")
-            .flag("-march=sapphirerapids")
-            .flag("-mavx512f")
-            .flag("-mavx512vl")
-            .flag("-mavx512bw")
-            .flag("-mavx512vnni")
-            .flag("-mavx512fp16")
-            .flag("-O3")
-            .flag("-Wall")
-            .flag("-Werror")
-            .flag("-Wextra")
-            .flag("-Wpedantic")
-            .compile("f16");
+    if cfg!(all(target_arch = "aarch64", target_os = "macos")) {
+        // Build a version with NEON
+        build_f16_with_flags("neon", &["-mtune=apple-m1"]).unwrap();
+    } else if cfg!(all(target_arch = "aarch64", target_os = "linux")) {
+        // Build a version with NEON
+        build_f16_with_flags("neon", &["-march=armv8.2-a+fp16"]).unwrap();
     }
+
+    if cfg!(target_arch = "x86_64") {
+        // Build a version with AVX512
+        if let Err(err) = build_f16_with_flags("avx512", &["-march=sapphirerapids"]) {
+            // It's likely the compiler doesn't support the sapphirerapids architecture
+            // Clang 12 and GCC 11 are the first versions with sapphire rapids support
+            eprintln!("Skipping Sapphire Rapids build due to error: {}", err);
+        } else {
+            // We create a special cfg so that we can detect we have in fact
+            // generated the AVX512 version of the f16 kernels.
+            println!("cargo:rustc-cfg=kernel_suppport=\"avx512\"");
+        };
+        // Build a version with AVX
+        // While GCC doesn't have support for _Float16 until GCC 12, clang
+        // has support for __fp16 going back to at least clang 6.
```

Comment:
is there a reason to use broadwell instead of haswell? 
Haswell is the oldest CPUs supported on AWS / GCP

Reply:
I can switch this to haswell